docs/ImageInput.md

```diff
@@ -170,3 +170,214 @@ The `<ImageInput>` component accepts the usual `className` prop. You can also ov
+            const newPictures = params.data.pictures.filter(
+                p => p.rawFile instanceof File
+            );
+            const formerPictures = params.data.pictures.filter(
+                p => !(p.rawFile instanceof File)
+            );
+
+            const base64Pictures = await Promise.all(
+                newPictures.map(convertFileToBase64)
+            )
+            const pictures = [
+                ...base64Pictures.map((dataUrl, index) => ({
+                    src: dataUrl,
+                    title: newPictures[index].name,
+                })),
+                ...formerPictures,
+            ];
+            return dataProvider.update(
+                resource,
+                { data: { ...params.data, pictures } }
+            );  
+        }
+    }
+]);
+
+/**
+ * Convert a `File` object returned by the upload input into a base 64 string.
+ * That's not the most optimized way to store images in production, but it's
+ * enough to illustrate the idea of dataprovider decoration.
+ */
+const convertFileToBase64 = file =>
+    new Promise((resolve, reject) => {
+        const reader = new FileReader();
+        reader.onload = () => resolve(reader.result);
+        reader.onerror = reject;
+        reader.readAsDataURL(file.rawFile);
+    });
+
+export default myDataProvider;
+```
+
+**Tip**: Use `beforeSave` instead of `beforeUpdate` to do the same for both create and update calls.
+
+### How To Send File Using FormData
+
+In case you need to upload files to your API, as you would with an HTML form, you can use [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) API that uses the same format a form would use if the encoding type were set to `multipart/form-data`.
+
+The following `dataprovider` extends an existing one and leverages [`withLifecycleCallbacks`](#adding-lifecycle-callbacks) to modify the `dataProvider.create()` and `dataProvider.update()` methods for the `posts` resource only with the [`beforeSave`](./withLifecycleCallbacks.md#beforesave) methods. 
+
+It creates a new `FormData` object with the file received from the form and sends this file to the API. It is the role of your API to negotiate the request and process the image.
+
+Then it waits for the API response and fills the `params.picture` object with the image source and title sent by the API. The `params` object is finally returned to allow the parent DataProvider to carry out its processes.
+
+```ts
+import { DataProvider, withLifecycleCallbacks } from 'react-admin';
+import simpleRestProvider from 'ra-data-simple-rest';
+
+const dataProvider = withLifecycleCallbacks(simpleRestProvider('http://path.to.my.api/'), [
+  {
+    resource: "posts",
+    beforeSave: async (params: any, dataProvider: DataProvider) => {
+      const formData = new FormData();
+      formData.append("file", params.picture.rawFile);
+  
+      const imageResponse = await fetch(`http://path.to.my.api/posts`, {
```

Comment:
Isn't it unlikely that the URL for image upload is the same as the post update URL?

Reply:
I thought the API could detect if a file is present on the request, but yes, it will be clearer if the upload URL is different.